[IOSP-514] [CNSMR-3272] Technical Documentation for Stevenson and CRP (#46)
Babylonpartners/Stevenson: Documentation/BabylonCommands.md

* Invoking a `fastlane` command on CI
* Responding to a GitHub comment to trigger a CI pipeline
* [Generating the CRP ticket](CRP.md)

## `/stevenson`

This is the main Slack command, used to namespace other commands. It's implemented in `MainCommand.swift`

This command expects one of the following sub-commands, which are described in more details below:

* `/stevenson fastlane …`
* `/stevenson appcenter …`
* `/stevenson testflight …`
* `/stevenson crp …`

## `/stevenson fastlane`

This and the 2 following commands are implemented in `SlackCommand+Fastlane.swift`
> Also still accessible via `/fastlane` directly, though deprecated in favor of being invoked as a `/stevenson` subcommand instead.

This command allows you to trigger a CircleCI job which will run the provided lane.

```
/stevenson fastlane <lane> [branch:<branch>] [target:<target>] [<other_options>]
```

* If `branch:` is not provided, defaults to running the lane on `develop`
* `target:` is only used by some lanes in Fastlane, like `appcenter` or `testflight`. In most cases it's not needed

> * Any potential `<other_options>` depend on each lane. Check our Fastlane to see which are available for each. Examples include `version:` for `appcenter` and `testflight` lanes

Reviewer: full examples would help understanding the command. Especially for people unfamiliar with this kind of flow.
I think full examples would be useful for commands below too

Author: I didn't want those to be too much detailed in here – as I was trying to focus on technical details and implementation, which any _usage documentation_ of babylon-specific commands should be documented in the playbook, not here.

tbh I was even hesitant to keep that `BabylonCommands.md` document here in the repo – its goal was just to describe in a technical POV examples of commands that we could (and did) implement with the `Stevenson.framework`, for anyone else in the OSS community that would be interested to use those as examples/inspiration to implement their own bot using Stevenson-the-framework

I could totally add examples in here but I feel it would make more sense to keep detailed usage documentation in our playbook. Maybe we could even try to remove some paragraphs in that `BabylonCommands.md` doc to really only focus on the implementation details rather than usage description, but it's hard to talk about implementation without explaining even a little the expected thing we are trying to parse 😉

Reviewer: I thought this work was strictly about the process ticket CNSMR-3272. That doc should focus more on the usage rather then implementation. It would be great if you can keep this document that you have just created and have a shorter version focused on the usage. What do you think?

Author: Yeah the separation as to where document things is not always clear indeed.

We agreed that the documentation should ideally be split in:
 - General Technical documentation about the framework itself (intended for users of the OSS framework) ➡️ in the repo itself
 - Anything specific to Babylon (especially process and specific commands) ➡️ in the playbook

Maybe I can just move this `BabylonCommands.md` doc in the playbook after all then since it's about the usage and specific to commands implemented for Babylon (and not doc for the bot framework intended for OSS users). But **we already have documentation in the playbook describing those commands [here](https://github.com/babylonhealth/ios-playbook/blob/master/Cookbook/Technical-Documents/SlackCIIntegration.md) though** (as well as in the playbook doc about release process) so it would not make sense to duplicate the documentation.

Maybe I will just delete that `BabylonCommands.md` doc entirely then since it doesn't really talk about implementation details after all (and usage doc already exists in playbook), wdyt?

Author: @viorel15 as discussed in Slack, cleaned up the `BabylonCommands.md` to only focus on implementation details, and links to usage in playbook otherwise.

Still kept the document since we needed a place especially to document the environment variables used by our bot instance (being technical documentation for the bot, I believe that belonged more in this repo than the playbook), but otherwise simplified the doc to only point to relevant places.

Reviewer 2: @AliSoftware if that's the only purpose of this doc that's left (to me it feels it does not add much more clarity about implementation details that we already have in other docs or in code except describing the structure of the project) wouldn't be better to document environment variables in code instead of a separate document and then remove this doc? We can have a small section in the main README dedicated to the structure of the project and references that we have in a playbook.

```
/stevenson appcenter <target> [version:<version>] [branch:<branch>]
```

* This command is just a convenience doing the same as `/stevenson fastlane appcenter target:<target> version:<version> branch:<branch>`
* If `branch:` is not provided, defaults to `release/<target>/<version>`

## `/stevenson testflight`

> Also still accessible via `/testflight` directly, though deprecated in favor of being invoked as a `/stevenson` subcommand instead.

```
/stevenson testflight <target> [version:<version>] [branch:<branch>]
```

* This command is just a convenience doing the same as `/stevenson fastlane testflight target:<target> version:<version> branch:<branch>`
* If `branch:` is not provided, defaults to `release/<target>/<version>`

## `/stevenson crp`

> Still accessible via `/crp` directly, though deprecated in favor of new name.

This command creates a "CRP Ticket" in our JIRA, which is a JIRA ticket gathering all the tickets that will be part of a release to the Stores in order to validate approval for that release. The creation and approval of this ticket to make a new release is part of our SSDLC. 

> For more information about the CRP, visit the dedicated page in your Playbook.

Reviewer: Can you please link the page?

Author: Sure 👍 

It was not created when I wrote that PR, but now the PR to create the page exists in playbook (https://github.com/babylonhealth/ios-playbook/pull/313) so I should be able to link it

```suggestion
For more information about the CRP, visit [the dedicated page in your Playbook](https://github.com/babylonhealth/ios-playbook/blob/master/Cookbook/Technical-Documents/CRP-Bot.md).
```
_(link will obviously not work until ios-playbook#313 is merged to master)_

> For more information about the CRP, visit [the dedicated page in your Playbook](https://github.com/babylonhealth/ios-playbook/blob/master/Cookbook/Technical-Documents/CRP-Bot.md).

Author: PR pending for this future document: https://github.com/babylonhealth/ios-playbook/pull/313
(So not yet available in master… until the PR in playbook is merged)

Link to current version in branch in the meantime: https://github.com/babylonhealth/ios-playbook/blob/olivier/IOSP-514/Cookbook/Technical-Documents/CRP-Bot.md

```
/crp <repo> branch:<branch>
```

* `<repo>` should be one of the repositories declared in `RepoMapping.swift` – namely either `ios` or `android`.
* `<branch>` will typically be the name of a release branch, e.g. `release/babylon/<version>`

Execution of CRP process is composed of multiple subtasks (auto-gathering the list of tickets from the commits, creating the CRP ticket, creating JIRA versions on each relevant boards, setting the "Fix Version" field of all relevant tickets to the appropriate JIRA version…) and involves a good chunk of code. For implementation details of this part of the code, see [CRP implementation details](CRP.md)

## Responding to a GitHub comment `"@ios-bot-babylon …"`

Our bot also responds to a route to the `/github/comment` URL, which is sent by GitHub webhooks when a new comment is made on a GitHub issue or Pull Request.

We have configured our GitHub repo to send the webhook to the URL of our bot, which means our bot will handle every new comment made to any issue or PR and be able to act on it.

> This route processes comments starting with `@ios-babylon-bot` (discarding any other kind of comments); it then runs `fastlane` to invoke the mentioned lane on the PR's branch in CI.

Author: Actually it invokes a CI workflow, not a fastlane lane.
```suggestion
This route processes comments starting with `@ios-babylon-bot` (discarding any other kind of comments); it then invokes the workflow (whose name is provided in the comment) on the PR's branch in CircleCI.
```

```

> @ios-bot-babylon <lane_name>

Author: ```suggestion
@ios-bot-babylon <workflow_name>
```

```

> This will simply run lane `<lane_name>` in CI on the head branch of the PR this comment was made on.

Author: ```suggestion
This will simply run the workflow `<workflow_name>` on CircleCI, on the head branch of the PR this comment was made on.
```

# Babylon's Stevenson Commands

> This Vapor project consists actually of two projects:

Reviewer 3: [nit]
```suggestion
This Vapor project actually consists of two projects:
```

 * The Stevenson framework, aimed to be reused so you can create your own bots interacting with Slack, GitHub, CircleCI and JIRA yourselves

>  * The Stevenson Babylon app, which is the instance of the Bot we use at Babylon, based on the Stevenson framework, and which provides us with the Slack commands we need

Reviewer 2: ```suggestion
 * The Stevenson Babylon app, which is the instance of the Bot we use at Babylon, based on the Stevenson framework, and which provides us with commands we need
```

# ⚙️ Environment Variables used by our app

The bot instance of Stevenson we use at Babylon uses environment variables to instantiate and configure the various services (see [README.md](../README.md#supported-services)).

See `app.swift` for declaration of those environment variables used in the app, and `configure.swift` for where they are used to instantiate the services.

ENV Var | Description
--------|--------------

> `SLACK_TOKEN` | Your access token for the Slack API of your team's Slack

Reviewer 2: imo it would be better to comment that in code, if the purpose of these environment variables is not clear, not to go out of sync. Readme already mentions environment variables so I'd leave this whole section out.

`GITHUB_USERNAME`<br/>`GITHUB_TOKEN`| Login and access token of a user having read access to those repositories
`CIRCLECI_TOKEN`| Your access token for the CircleCI API
`JIRA_BASEURL`| Host name of your Jira instance (e.g. `https://myorg.atlassian.net:443`)
`JIRA_USERNAME`<br/>`JIRA_TOKEN`| Login and access token for the API of your Jira instance

# 🕹 Commands implemented in the Babylon Bot

> The Stevenson instance used at Babylon, whose code can be found in the `App/` directory, mainly consist of 3 main core features

Reviewer 3: ```suggestion
The Stevenson instance used at Babylon, whose code can be found in the `App/` directory, consist of 3 main core features
```

## Invoking a `fastlane` command on CI via Slack

Allows us to use a slash command in our Slack to trigger CI workflows.

* Usage and supported commands are documented [in our playbook here](https://github.com/babylonhealth/ios-playbook/blob/master/Cookbook/Technical-Documents/SlackCIIntegration.md).

> * The commands are implemented/defined in `SlackCommand+Fastlane.swift` and then handled in `SlackService.swift` by `func handle(command:on:)`.

Reviewer 2: This is pretty clear from code already, isn't it?

```suggestion
* The commands are implemented/defined in `SlackCommand+Fastlane.swift`.
```

> Implementation simply consists of parsing the slack command parameters, then invoking `CircleCIService.runlane` with the right arguments.

Reviewer 2: Is this really helpful and not already clear from code? I'd also avoid using "simply"/"just"/etc.

```suggestion
Implementation consists of parsing the slack command parameters, then invoking `CircleCIService.runlane` with the right arguments.
```

Our Slack is configured to send webhooks for those slash commands to the URL of our bot for it to process those requests – See [main README.md](../README.md) for details.

## Responding to a GitHub comment to trigger a CI pipeline

* Usage documentation can be found [here](https://github.com/babylonhealth/ios-playbook/blob/master/Cookbook/Technical-Documents/SlackCIIntegration.md)

> * Implementation simply consists of a _route_ implemented in `routes.swift` and invoking `GitHubService.issueComment(on:ci:)` (in `GitHubService+IssueComment.swift`).  

Reviewer 2: IMO that's also pretty clear from the code. What we can mention here is that comments handling is implemented as a separate route in the app.

```suggestion
* Implementation consists of a _route_ implemented in `routes.swift` and invoking `GitHubService.issueComment(on:ci:)` (in `GitHubService+IssueComment.swift`).  
```

>   The method parses the webhook request headers and the comment body, then calls `CircleCIService.runLane` or `CircleCIService.runPipeline` appropriately to trigger the right CI workflow.

Reviewer 2: ```suggestion
  The method parses the webhook request headers and the comment body and then triggers appropriate CI workflow (that runs a single lane or a pipeline).
```
